### PR TITLE
refactor interactions with multiplayer slots

### PR DESCRIPTION
**Pull request by cmyui**

### CODE REVIEW

File: objects/match.py [hunk 1]
```diff
@@ -807,7 +834,7 @@ def changeTeam(self, userID: int, newTeam: Optional[int] = None) -> None:
 
         # Update slot and send update
         if newTeam is None:
-            if self.slots[slotID].team:
+            if self.slots[slotID].team == matchTeams.RED:
```
Comment (cmyui) [inline objects/match.py:842]: this fixes [AKAT-184](https://osuakatsuki.atlassian.net/browse/AKAT-184) - Unable to toggle from blue to red team

**Review by infernalfire72 — DISMISSED** [review]

File: objects/match.py [hunk 2]
```diff
@@ -153,34 +154,42 @@ def getMatchData(self, censored: bool = False) -> tuple[tuple[object, int]]:
...
+        old_host = glob.tokens.getTokenFromUserID(self.hostUserID)
+        assert old_host is not None
 
-        slotID = self.getUserSlotID(newHost)
-        if (
-            slotID is None or
-            self.slots[slotID].user not in glob.tokens.tokens
-        ):
+        if not old_host.staff:
+            self.remove_referee(self.hostUserID)
```
Comment (cmyui) [inline objects/match.py:177]: this fixes [AKAT-185](https://osuakatsuki.atlassian.net/browse/AKAT-185) - Passing host off while being staff will remove referee status

File: objects/match.py [hunk 3]
```diff
@@ -968,5 +1001,6 @@ def __enter__(self) -> 'match':
 
     def __exit__(self, exc_type: Optional[Type[BaseException]],
                  exc_value: Optional[BaseException],
-                 traceback: Optional[TracebackType]) -> None:
+                 traceback: Optional[TracebackType]) -> bool:
         self._lock.release()
+        return exc_value is not None
```
Comment (tsunyoku) [inline objects/match.py]: this is buggy behaviour from testing, wouldn't recommend

**Review by tsunyoku — CHANGES_REQUESTED** [review]

File: objects/match.py [hunk 4]
```diff
@@ -968,5 +1006,6 @@ def __enter__(self) -> 'match':
 
     def __exit__(self, exc_type: Optional[Type[BaseException]],
                  exc_value: Optional[BaseException],
-                 traceback: Optional[TracebackType]) -> None:
+                 traceback: Optional[TracebackType]) -> bool:
```
Comment (tsunyoku) [inline objects/match.py]:
```suggestion
                 traceback: Optional[TracebackType]) -> None:
```

File: objects/match.py [hunk 5]
```diff
@@ -968,5 +1006,6 @@ def __enter__(self) -> 'match':
...
         self._lock.release()
+        return exc_value is not None
```
Comment (tsunyoku) [inline objects/match.py]:
```suggestion
```

**Review by tsunyoku — APPROVED** [review]

**Pull request merged**